File: JAVA_Templates/DSU.java
package JAVA_Templates;

class DSU {
    int[] Parent, Group_Size;
    int Number_of_Nodes, Number_of_Groups, Max_Group;

    public DSU(int Number_of_Nodes) {
        this.Number_of_Nodes = Number_of_Nodes;
        Parent = new int[Number_of_Nodes + 1];
        Group_Size = new int[Number_of_Nodes + 1];
        Number_of_Groups = Number_of_Nodes;
        Max_Group = 1;
        for (int i = 1; i <= Number_of_Nodes; i++) {
            Parent[i] = i;
            Group_Size[i] = 1;
        }
    }

    public int Leader(int x) {
        return Parent[x] = (Parent[x] == x ? x : Leader(Parent[x]));
    }

    public boolean Is_same_Group(int x, int y) {
        return Leader(x) == Leader(y);
    }

    public void unite(int x, int y) {
        int leader1 = Leader(x);
        int leader2 = Leader(y);
        if (leader1 != leader2) {
            Number_of_Groups--;
            if (Group_Size[leader1] < Group_Size[leader2]) {
                int temp = leader1;
                leader1 = leader2;
                leader2 = temp;
            }
            Parent[leader2] = leader1;
            Group_Size[leader1] += Group_Size[leader2];
            Max_Group = Math.max(Max_Group, Group_Size[leader1]);
        }
    }

    public int getSize(int x) {
        return Group_Size[Leader(x)];
    }
}

File: python_templates/dsu.py
class DSU:
    def __init__(self, n: int):
        self.n = n
        self.parent = [0] * (n + 1)
        self.group_size = [1] * (n + 1)
        self.number_of_nodes = n
        self.number_of_groups = n
        self.max_group = 1
        for i in range(1, n + 1):
            self.parent[i] = i

    def leader(self, x: int) -> int:
        if self.parent[x] == x:
            return x
        self.parent[x] = self.leader(self.parent[x])
        return self.parent[x]

    def is_same_group(self, u: int, v: int) -> bool:
        return self.leader(u) == self.leader(v)

    def unit(self, u: int, v: int) -> None:
        leader1 = self.leader(u)
        leader2 = self.leader(v)
        if leader1 != leader2:
            self.number_of_groups -= 1
            if self.group_size[leader1] < self.group_size[leader2]:
                temp = leader1
                leader1 = leader2
                leader2 = temp
            self.parent[leader2] = leader1
            self.group_size[leader1] += self.group_size[leader2]
            self.max_group = max(self.max_group, self.group_size[leader1])

    def get_size(self, x: int) -> int:
        return self.group_size[self.leader(x)]
